docs/filezilla.man
.TH filezilla 1 "January 2018" "" "FileZilla Manual"
.SH NAME
FileZilla \- FTP client
.SH SYNOPSIS
.B filezilla
.br
.B filezilla [\-l <logontype>] <FTP URL>
.br
.B filezilla \-h|\-s|\-v
.br
.B filezilla \-c <site>
.br
.B filezilla \-a <path>
.SH DESCRIPTION
FileZilla is a powerful client for plain FTP, FTP over SSL/TLS (FTPS) and the SSH File Transfer Protocol (SFTP).
.SH OPTIONS

.TP
.B \-c <site>, \-\-site <site>
Connect to the given site from the Site Manager.
Site has to be given as complete path, with a slash as separation character. Any slash or backslash that is part of a segment has to be escaped with a backslash. Path has to be prefixed with 0 for user defined entries or 1 for default entries. Site path may not contain double quotation marks.
.IP
Example: filezilla \-c 0/foo/bar/sl\\/ash connects to the user site sl/ash in the site directory foo/bar
.IP
May not be used together with
.B \-s
nor with URL parameter.

.TP
.B \-a <path>, \-\-local <path>
Sets the local site to the given path.
.IP
Example: filezilla \-a "/foo bar/"

.TP
.B \-h, \-\-help
Displays a help dialog listing these commandline options.

.TP
.B \-l <logontype>, \-\-logontype <logontype>
Set a special logontype, can only be used in combination with a FTP URL as argument.

Logontype has to be either
.B ask
or
.BR interactive ". If " \-l " isn't given, the normal logontype is used."

.TP
.B \-s, \-\-sitemanager
Start with Site Manager opened. May not be used together with
.B \-c
nor with URL parameter.

.TP
.B \-v, \-\-version
Display version number of FileZilla.


.SH ENVIRONMENT

The
.B FZ_DATADIR
environment variable can be used to specify the directory containing FileZilla's data files.
.SH SUPPORT
Please visit https://filezilla\-project.org/ for further information. Report bugs only if you are using the latest version available from the FileZilla website.
.SH COPYRIGHT
Copyright (C) 2004\-2019  Tim Kosse
.P
FileZilla is distributed under the terms of the GNU General Public License version 2 or later.
.SH "SEE ALSO"
fzdefaults.xml(5)


NEWS
This file only lists a short summary of the changes between FileZilla versions. For a full list of changes, please read the ChangeLog file.

Legend:
  ! Security related bugfix or otherwise critical bugfix
  + New feature or important bugfix
  - Bugfixes and other changes

  Platform prefixes:
    MSW: Microsoft Windows
    *nix: Unix (e.g BSD) and Unix-like systems (e.g. Linux)
    macOS/OS X: Apple macOS

3.55.0 (2021-07-09)

- Updated translations

3.55.0-rc1 (2021-07-02)

- Updated SFTP components
- Set appropriate ALPN on TLS connections
- Fixed loading of settings files that were moved between platforms
- Refactored updater internals

3.54.1 (2021-05-13)

- Fix control visibility logic in Site Manager

3.54.0 (2021-05-12)

- SFTP: Fixed a regression loading configured key files

3.54.0-rc1 (2021-05-04)

- Refactored network activity handling, the activity indicators tooltips now display more accurate data
- Updated SFTP components
- Small performance improvement sorting very large directory listings
- Fixed a crash changing numeric permissions value in chmod dialog

3.53.1 (2021-03-22)

- Removed an error message on new installation where there is no sitemanager.xml yet
- Fixed SOCKS5 proxy support

3.53.0 (2021-03-15)

- HTTP: Handle servers sending early replies

3.53.0-rc1 (2021-03-05)

+ Require explicit confirmation on FTP servers not supporting TLS session esumption on the data connection
+ Access Grant support for Tardigrade
- MSW: Fixed handle leak if using SFTP
- FTP: Fixed cancelling listings on very slow computers with extremely fast network connection
- Fixed loading of custom FTP proxy setting when opening the settings dialog
- Internal restructuring

3.52.2 (2021-01-18)

- When restoring the main window position at startup, detect if the title bar would become hidden behind taskbars or global menu bars and reposition it appropriately
- SFTP: Fixed preserving of timestamps on downloads

3.52.0.5 (2021-01-08)

- FTP: Fixed crash if downloading ASCII files with stray carriage returns
- Fixed too strict requirements for preallocation
- *nix: Fixed a memory leak

3.52.0.4 (2021-01-07)

- macOS: Work around a kernel bug where ftruncate cannot be called more than once on a shared memory object

3.52.0.3 (2021-01-06)

- macOS: Fix issue opening shared memory objects

3.52.0.2 (2021-01-06)

- HTTP: Handle multi-token Connection: headers
	
3.52.0.1 (2021-01-05)

- macOS: Rebuilt to address a toolchain issue

3.52.0 (2021-01-04)

- Fixed a regression with the rename file exists action on downloads

3.52.0-rc1 (2020-12-29)

+ Added context menu item in the search dialog to close the search and set the selected directory as the current local/remote directory
- Minor fixes for updating the HTTP transfer status

3.52.0-beta1 (2020-12-17)

+ Use asynchronous file readers/writers in the engine
- MSW: React to system color scheme changes
- Fixed splitters not being draggable on newly created tabs
- MSW: Fixed data directory lookup if run from a development environment

3.51.0 (2020-10-20)

- Fixed import of sites protected with a different master password
- Directory comparison now works correctly with non-default file name sorting settings
- Small search dialog fixes

3.51.0-rc1 (2020-10-13)

- Refactored how the queue stores transfer flags
- Refactored internals to be more modular
- Linux: Disable receive buffer tuning if it results in unexpected consequences
- Distinguish between Cancel and No in the file changed dialog
- Small fixes to changing or disabling a master password

3.50.0 (2020-08-27)

- Minor fixes to Tardigrade integration

3.50.0-rc1 (2020-08-21)

+ Update Storj/Tardigrade integration, now based on the uplink-c library
- Fix layout issues in the Site Manager if switching protocol
- Restored natural sort option for file list sorting

3.49.1 (2020-07-15)

- Fixed selection changes not being reflected in the Site Manager in some circumstances

3.49.0 (2020-07-13)

- Fixed retrying initial connection attempts

3.49.0-rc1 (2020-07-07)

- MSW: Tab bar icons now scale on high-DPI displays
- The sizing and spacing of several dialogs has been improved on high-DPI displays
- MSW: Fix visual glitches when changing the selected site in the Site Manager
- When creating a new tab, the pane sizes are now inherited from the previous tab.
- macOS: Additional fixes for handling Cmd+V, Cmd+C and Cmd+X in text input controls
- Refactored internal settings and option handling

3.48.1 (2020-05-18)

- If FileZilla is configured to restore open tabs on startup, the directory trees no longer shrink in size

3.48.1-rc1 (2020-05-12)

! MSW: The settings directory is now initially created with tight permissions restricted to the user's account. Previously the directory could have had inherited too wide permissions if a custom location outside the user's profile directory had been preconfigured.
+ Redesigned certificate dialog
+ During directory comparison and comparative search, Ctrl+Shift+A can now be used to select only the highlighted files
- Fixed import of settings
- Fixed local file editing if the option to track local file changes had been disabled
- Fixes to storing associations

3.48.0 (2020-04-27)

! Custom associations for file extensions containing spaces were not saved correctly

3.48.0-rc1 (2020-04-20)

+ Overhaul of file associations for editing files. The setting to inherit system associations has been removed. When first editing a type of file with no default editor set, a dialog is shown asking which editor to use. The system associatio, if it exists, is a choice in this dialog.
- SFTP: Re-added support for AES-GCM which was accidentally removed in 3.46.1
- Fixes for comparative search leaving the program in an inconsistent state when closing the dialog during an ongoing search

src/interface/FileZilla.cpp
#include "filezilla.h"
#ifdef _MSC_VER
#pragma hdrstop
#endif
#include "filezillaapp.h"
#include "Mainfrm.h"
#include "Options.h"
#include "wrapengine.h"
#include "buildinfo.h"
#include "cmdline.h"
#include "welcome_dialog.h"
#include "msgbox.h"
#include "themeprovider.h"
#include "wxfilesystem_blob_handler.h"
#include "../commonui/fz_paths.h"
#include "../include/version.h"
#include <libfilezilla/local_filesys.hpp>
#include <libfilezilla/translate.hpp>

#include <wx/evtloop.h>

#ifdef WITH_LIBDBUS
#include "../dbus/session_manager.h"
#endif

#if defined(__WXMAC__) || defined(__UNIX__)
#include <wx/stdpaths.h>
#elif defined(__WXMSW__)
#include <shobjidl.h>
#endif

#include "locale_initializer.h"

#ifdef USE_MAC_SANDBOX
#include "osx_sandbox_userdirs.h"
#endif

#if !defined(__WXGTK__) && !defined(__MINGW32__)
IMPLEMENT_APP(CFileZillaApp)
#else
IMPLEMENT_APP_NO_MAIN(CFileZillaApp)
#endif //__WXGTK__

#if !wxUSE_PRINTF_POS_PARAMS
  #error Please build wxWidgets with support for positional arguments.
#endif

namespace {
#if FZ_WINDOWS
std::wstring const PATH_SEP = L";";
#define L_DIR_SEP L"\\"
#else
std::wstring const PATH_SEP = L":";
#define L_DIR_SEP L"/"
#endif
}

CFileZillaApp::CFileZillaApp()
{
	m_profile_start = fz::monotonic_clock::now();
	AddStartupProfileRecord("CFileZillaApp::CFileZillaApp()");
}

CFileZillaApp::~CFileZillaApp()
{
	themeProvider_.reset();
	COptions::Destroy();
}

void CFileZillaApp::InitLocale()
{
	wxString language = COptions::Get()->get_string(OPTION_LANGUAGE);
	const wxLanguageInfo* pInfo = wxLocale::FindLanguageInfo(language);
	if (!language.empty()) {
#ifdef __WXGTK__
		if (CInitializer::error) {
			wxString error;

			wxLocale *loc = wxGetLocale();
			const wxLanguageInfo* currentInfo = loc ? loc->GetLanguageInfo(loc->GetLanguage()) : 0;
			if (!loc || !currentInfo) {
				if (!pInfo) {
					error.Printf(_("Failed to set language to %s, using default system language."),
						language);
				}
				else {
					error.Printf(_("Failed to set language to %s (%s), using default system language."),
						pInfo->Description, language);
				}
			}
			else {
				wxString currentName = currentInfo->CanonicalName;

				if (!pInfo) {
					error.Printf(_("Failed to set language to %s, using default system language (%s, %s)."),
						language, loc->GetLocale(),
						currentName);
				}
				else {
					error.Printf(_("Failed to set language to %s (%s), using default system language (%s, %s)."),
						pInfo->Description, language, loc->GetLocale(),
						currentName);
				}
			}

			error += _T("\n");
			error += _("Please make sure the requested locale is installed on your system.");
			wxMessageBoxEx(error, _("Failed to change language"), wxICON_EXCLAMATION);

			COptions::Get()->set(OPTION_LANGUAGE, _T(""));
		}
#else
		if (!pInfo || !SetLocale(pInfo->Language)) {
			for (language = GetFallbackLocale(language); !language.empty(); language = GetFallbackLocale(language)) {
				const wxLanguageInfo* fallbackInfo = wxLocale::FindLanguageInfo(language);
				if (fallbackInfo && SetLocale(fallbackInfo->Language)) {
					COptions::Get()->set(OPTION_LANGUAGE, language.ToStdWstring());
					return;
				}
			}
			COptions::Get()->set(OPTION_LANGUAGE, std::wstring());
			if (pInfo && !pInfo->Description.empty()) {
				wxMessageBoxEx(wxString::Format(_("Failed to set language to %s (%s), using default system language"), pInfo->Description, language), _("Failed to change language"), wxICON_EXCLAMATION);
			}
			else {
				wxMessageBoxEx(wxString::Format(_("Failed to set language to %s, using default system language"), language), _("Failed to change language"), wxICON_EXCLAMATION);
			}
		}
#endif
	}
}

namespace {
std::wstring translator(char const* const t)
{
	return wxGetTranslation(t).ToStdWstring();
}

std::wstring translator_pf(char const* const singular, char const* const plural, int64_t n)
{
	// wxGetTranslation does not support 64bit ints on 32bit systems.
	if (n < 0) {
		n = -n;
	}
	return wxGetTranslation(singular, plural, (sizeof(unsigned int) < 8 && n > 1000000000) ? (1000000000 + n % 1000000000) : n).ToStdWstring();
}
}

bool CFileZillaApp::OnInit()
{
	AddStartupProfileRecord("CFileZillaApp::OnInit()");

	// Turn off idle events, we don't need them
	wxIdleEvent::SetMode(wxIDLE_PROCESS_SPECIFIED);

	wxUpdateUIEvent::SetMode(wxUPDATE_UI_PROCESS_SPECIFIED);

	fz::set_translators(translator, translator_pf);

#ifdef __WXMSW__
	SetCurrentProcessExplicitAppUserModelID(L"FileZilla.Client.AppID");
#endif

	//wxSystemOptions is slow, if a value is not set, it keeps querying the environment
	//each and every time...
	wxSystemOptions::SetOption(_T("filesys.no-mimetypesmanager"), 0);
	wxSystemOptions::SetOption(_T("window-default-variant"), _T(""));
#ifdef __WXMSW__
	wxSystemOptions::SetOption(_T("no-maskblt"), 0);
	wxSystemOptions::SetOption(_T("msw.window.no-clip-children"), 0);
	wxSystemOptions::SetOption(_T("msw.font.no-proof-quality"), 0);
	wxSystemOptions::SetOption(_T("msw.remap"), 0);
	wxSystemOptions::SetOption(_T("msw.staticbox.optimized-paint"), 0);
#endif
#ifdef __WXMAC__
	wxSystemOptions::SetOption(_T("mac.listctrl.always_use_generic"), 1);
	wxSystemOptions::SetOption(_T("mac.textcontrol-use-spell-checker"), 0);
#endif

	int cmdline_result = ProcessCommandLine();
	if (!cmdline_result) {
		return false;
	}

	LoadLocales();

	if (cmdline_result < 0) {
		if (m_pCommandLine) {
			m_pCommandLine->DisplayUsage();
		}
		return false;
	}

#if USE_MAC_SANDBOX
	// Set PUTTYDIR so that fzsftp uses the sandboxed home to put settings.
	std::wstring home = GetEnv("HOME");
	if (!home.empty()) {
		if (home.back() != '/') {
			home += '/';
		}
		wxSetEnv("PUTTYDIR", home + L".config/putty");
	}
#endif

	COptions::Init();

	InitLocale();

#ifndef _DEBUG
	const wxString& buildType = CBuildInfo::GetBuildType();
	if (buildType == _T("nightly")) {
		wxMessageBoxEx(_T("You are using a nightly development version of FileZilla 3, do not expect anything to work.\r\nPlease use the official releases instead.\r\n\r\n\
Unless explicitly instructed otherwise,\r\n\
DO NOT post bugreports,\r\n\
DO NOT use it in production environments,\r\n\
DO NOT distribute the binaries,\r\n\
DO NOT complain about it\r\n\
USE AT OWN RISK"), _T("Important Information"));
	}
	else {
		std::wstring v = GetEnv("FZDEBUG");
		if (v != L"1") {
			COptions::Get()->set(OPTION_LOGGING_DEBUGLEVEL, 0);
			COptions::Get()->set(OPTION_LOGGING_RAWLISTING, 0);
		}
	}
#endif

	if (!LoadResourceFiles()) {
		COptions::Destroy();
		return false;
	}

	themeProvider_ = std::make_unique<CThemeProvider>();
	CheckExistsFzsftp();
#if ENABLE_STORJ
	CheckExistsFzstorj();
#endif

#ifdef WITH_LIBDBUS
	CSessionManager::Init();
#endif

	// Load the text wrapping engine
	m_pWrapEngine = std::make_unique<CWrapEngine>();
	m_pWrapEngine->LoadCache();

	bool welcome_skip = false;
#ifdef USE_MAC_SANDBOX
	OSXSandboxUserdirs::Get().Load();

	if (OSXSandboxUserdirs::Get().GetDirs().empty()) {
		CWelcomeDialog welcome;
		welcome.Run(nullptr, false);
		welcome_skip = true;

		OSXSandboxUserdirsDialog dlg;
		dlg.Run(nullptr, true);
		if (OSXSandboxUserdirs::Get().GetDirs().empty()) {
			return false;
		}
    }
#endif

	CMainFrame *frame = new CMainFrame();
	frame->Show(true);
	SetTopWindow(frame);

	if (!welcome_skip) {
		CWelcomeDialog::RunDelayed(frame);
	}

	frame->ProcessCommandLine();
	frame->PostInitialize();

	ShowStartupProfile();

	return true;
}

int CFileZillaApp::OnExit()
{
	CContextManager::Get()->NotifyGlobalHandlers(STATECHANGE_QUITNOW);

	COptions::Get()->Save();

#ifdef WITH_LIBDBUS
	CSessionManager::Uninit();
#endif

	if (GetMainLoop() && wxEventLoopBase::GetActive() != GetMainLoop()) {
		// There's open dialogs and such and we just have to quit _NOW_.
		// We cannot continue as wx proceeds to destroy all windows, which interacts
		// badly with nested event loops, virtually always causing a crash.
		// I very much prefer clean shutdown but it's impossible in this situation. Sad.
		_exit(0);
	}

	return wxApp::OnExit();
}

bool CFileZillaApp::LoadResourceFiles()
{
	AddStartupProfileRecord("CFileZillaApp::LoadResourceFiles");
	m_resourceDir = GetFZDataDir({L"resources/defaultfilters.xml"}, L"share/filezilla");

	wxImage::AddHandler(new wxPNGHandler());

	if (m_resourceDir.empty()) {
		wxString msg = _("Could not find the resource files for FileZilla, closing FileZilla.\nYou can specify the data directory of FileZilla by setting the FZ_DATADIR environment variable.");
		wxMessageBoxEx(msg, _("FileZilla Error"), wxOK | wxICON_ERROR);
		return false;
	}
	else {
		m_resourceDir.AddSegment(_T("resources"));
	}

	// Useful for XRC files with embedded image data.
	wxFileSystem::AddHandler(new wxFileSystemBlobHandler);

	return true;
}

bool CFileZillaApp::LoadLocales()
{
	AddStartupProfileRecord("CFileZillaApp::LoadLocales");
	m_localesDir = GetFZDataDir({L"locales/de/filezilla.mo"}, std::wstring());
	if (!m_localesDir.empty()) {
		m_localesDir.AddSegment(_T("locales"));
	}
#ifndef __WXMAC__
	else {
		m_localesDir = GetFZDataDir({L"de/filezilla.mo", L"de/LC_MESSAGES/filezilla.mo"}, L"share/locale", false);
	}
#endif
	if (!m_localesDir.empty()) {
		wxLocale::AddCatalogLookupPathPrefix(m_localesDir.GetPath());
	}

	SetLocale(wxLANGUAGE_DEFAULT);

	return true;
}

bool CFileZillaApp::SetLocale(int language)
{
	// First check if we can load the new locale
	auto pLocale = std::make_unique<wxLocale>();
	wxLogNull log;
	pLocale->Init(language);
	if (!pLocale->IsOk()) {
		return false;
	}

	// Load catalog. Only fail if a non-default language has been selected
	if (!pLocale->AddCatalog(_T("filezilla")) && language != wxLANGUAGE_DEFAULT) {
		return false;
	}
	pLocale->AddCatalog(_T("libfilezilla"));

	if (m_pLocale) {
		// Now unload old locale
		// We unload new locale as well, else the internal locale chain in wxWidgets get's broken.
		pLocale.reset();
		m_pLocale.reset();

		// Finally load new one
		pLocale = std::make_unique<wxLocale>();
		pLocale->Init(language);
		if (!pLocale->IsOk()) {
			return false;
		}
		else if (!pLocale->AddCatalog(_T("filezilla")) && language != wxLANGUAGE_DEFAULT) {
			return false;
		}
		pLocale->AddCatalog(_T("libfilezilla"));
	}
	m_pLocale = std::move(pLocale);

	return true;
}

int CFileZillaApp::GetCurrentLanguage() const
{
	if (!m_pLocale) {
		return wxLANGUAGE_ENGLISH;
	}

	return m_pLocale->GetLanguage();
}

wxString CFileZillaApp::GetCurrentLanguageCode() const
{
	if (!m_pLocale) {
		return wxString();
	}

	return m_pLocale->GetCanonicalName();
}

#if wxUSE_DEBUGREPORT && wxUSE_ON_FATAL_EXCEPTION
void CFileZillaApp::OnFatalException()
{
}
#endif

void CFileZillaApp::DisplayEncodingWarning()
{
	static bool displayedEncodingWarning = false;
	if (displayedEncodingWarning) {
		return;
	}

	displayedEncodingWarning = true;

	wxMessageBoxEx(_("A local filename could not be decoded.\nPlease make sure the LC_CTYPE (or LC_ALL) environment variable is set correctly.\nUnless you fix this problem, files might be missing in the file listings.\nNo further warning will be displayed this session."), _("Character encoding issue"), wxICON_EXCLAMATION);
}

CWrapEngine* CFileZillaApp::GetWrapEngine()
{
	return m_pWrapEngine.get();
}

void CFileZillaApp::CheckExistsFzsftp()
{
	AddStartupProfileRecord("FileZillaApp::CheckExistsFzsftp");
	CheckExistsTool(L"fzsftp", L"../putty/", "FZ_FZSFTP", OPTION_FZSFTP_EXECUTABLE, fztranslate("SFTP support"));
}

#if ENABLE_STORJ
void CFileZillaApp::CheckExistsFzstorj()
{
	AddStartupProfileRecord("FileZillaApp::CheckExistsFzstorj");
	CheckExistsTool(L"fzstorj", L"../storj/", "FZ_FZSTORJ", OPTION_FZSTORJ_EXECUTABLE, fztranslate("Storj support"));
}
#endif

void CFileZillaApp::CheckExistsTool(std::wstring const& tool, std::wstring const& buildRelPath, char const* env, engineOptions setting, std::wstring const& description)
{
	// Get the correct path to the specified tool

	bool found = false;
	std::wstring executable;

	std::wstring program = tool;
#ifdef __WXMAC__
	(void)buildRelPath;

	// On Mac we only look inside the bundle
	std::wstring path = GetOwnExecutableDir();
	if (!path.empty()) {
		executable = path + '/' + tool;
		if (FileExists(executable)) {
			found = true;
		}
	}
#else

#ifdef __WXMSW__
	program += L".exe";
#endif

	// First check the given environment variable
	executable = GetEnv(env);
	if (!executable.empty()) {
		if (FileExists(executable)) {
			found = true;
		}
	}

	if (!found) {
		std::wstring path = GetOwnExecutableDir();
		if (!path.empty()) {
			executable = path + program;
			if (FileExists(executable)) {
				found = true;
			}
			else {
				// Check if running from build dir
				if (path.size() > 7 && fz::ends_with(path, std::wstring(L"/.libs/"))) {
					if (FileExists(path.substr(0, path.size() - 6) + L"Makefile")) {
						executable = path + L"../" + buildRelPath + program;
						if (FileExists(executable)) {
							found = true;
						}
					}
				}
				else if (FileExists(path + L"Makefile")) {
					executable = path + buildRelPath + program;
					if (FileExists(executable)) {
						found = true;
					}
				}
			}
		}
	}

	if (!found) {
		// Check PATH
		std::wstring path = GetEnv("PATH");
		auto const segments = fz::strtok(path, PATH_SEP);
		for (auto const& segment : segments) {
			auto const cur = CLocalPath(segment).GetPath();
			executable = cur + program;
			if (!cur.empty() && FileExists(executable)) {
				found = true;
				break;
			}
		}
	}
#endif

	if (!found) {
		// Quote path if it contains spaces
		if (executable.find(' ') != std::wstring::npos && executable.front() != '"' && executable.front() != '\'') {
			executable = L"\"" + executable + L"\"";
		}

		wxMessageBoxEx(fz::sprintf(fztranslate("%s could not be found. Without this component of FileZilla, %s will not work.\n\nPossible solutions:\n- Make sure %s is in a directory listed in your PATH environment variable.\n- Set the full path to %s in the %s environment variable."), program, description, program, program, env),
			_("File not found"), wxICON_ERROR | wxOK);
		executable.clear();
	}
	COptions::Get()->set(setting, executable);
}

#ifdef __WXMSW__
extern "C" BOOL CALLBACK EnumWindowCallback(HWND hwnd, LPARAM)
{
	HWND child = FindWindowEx(hwnd, 0, 0, _T("FileZilla process identificator 3919DB0A-082D-4560-8E2F-381A35969FB4"));
	if (child) {
		::PostMessage(hwnd, WM_ENDSESSION, (WPARAM)TRUE, (LPARAM)ENDSESSION_LOGOFF);
	}

	return TRUE;
}
#endif

int CFileZillaApp::ProcessCommandLine()
{
	AddStartupProfileRecord("CFileZillaApp::ProcessCommandLine");
	m_pCommandLine = std::make_unique<CCommandLine>(argc, argv);
	int res = m_pCommandLine->Parse() ? 1 : -1;

	if (res > 0) {
		if (m_pCommandLine->HasSwitch(CCommandLine::close)) {
#ifdef __WXMSW__
			EnumWindows((WNDENUMPROC)EnumWindowCallback, 0);
#endif
			return 0;
		}

		if (m_pCommandLine->HasSwitch(CCommandLine::version)) {
			wxString out = wxString::Format(_T("FileZilla %s"), GetFileZillaVersion());
			if (!CBuildInfo::GetBuildType().empty()) {
				out += _T(" ") + CBuildInfo::GetBuildType() + _T(" build");
			}
			out += _T(", compiled on ") + CBuildInfo::GetBuildDateString();

			printf("%s\n", (const char*)out.mb_str());
			return 0;
		}
	}

	return res;
}

void CFileZillaApp::AddStartupProfileRecord(std::string const& msg)
{
	if (!m_profile_start) {
		return;
	}

	m_startupProfile.emplace_back(fz::monotonic_clock::now(), msg);
}

void CFileZillaApp::ShowStartupProfile()
{
	if (m_profile_start && m_pCommandLine && m_pCommandLine->HasSwitch(CCommandLine::debug_startup)) {
		AddStartupProfileRecord("CFileZillaApp::ShowStartupProfile");
		wxString msg = _T("Profile:\n");

		size_t const max_digits = fz::to_string((m_startupProfile.back().first - m_profile_start).get_milliseconds()).size();

		int64_t prev{};
		for (auto const& p : m_startupProfile) {
			auto const diff = p.first - m_profile_start;
			auto absolute = std::to_wstring(diff.get_milliseconds());
			if (absolute.size() < max_digits) {
				msg.append(max_digits - absolute.size(), wchar_t(0x2007)); // FIGURE SPACE
			}
			msg += absolute;
			msg += L" ";

			auto relative = std::to_wstring(diff.get_milliseconds() - prev);
			if (relative.size() < max_digits) {
				msg.append(max_digits - relative.size(), wchar_t(0x2007)); // FIGURE SPACE
			}
			msg += relative;
			msg += L" ";

			msg += fz::to_wstring(p.second);
			msg += L"\n";

			prev = diff.get_milliseconds();
		}
		wxMessageBoxEx(msg);
	}

	m_profile_start = fz::monotonic_clock();
	m_startupProfile.clear();
}

std::wstring CFileZillaApp::GetSettingsFile(std::wstring const& name) const
{
	return COptions::Get()->get_string(OPTION_DEFAULT_SETTINGSDIR) + name + _T(".xml");
}

#if defined(__MINGW32__)
extern "C" {
__declspec(dllexport) // This forces ld to not strip relocations so that ASLR can work on MSW.
int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR, int nCmdShow)
{
	wxDISABLE_DEBUG_SUPPORT();
	return wxEntry(hInstance, hPrevInstance, NULL, nCmdShow);
}
}
#endif

